File: MATLAB/example_26/VDM_model.m
function [X,Y,SS] = VDM_model(X,n,d,u0,ode_options,dt, ...
    t_end,mV,tbx)
% %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% %
% This function returns the simulated counts and summary metrics of the   %
% vandermeer (VDM) predator-prey model                                    % 
%                                                                         %
%  SYNOPSIS                                                               %
%   [X,Y,SS] = VDM_model(X,n,d,u0,ode_options,dt,t_end,mV,tbx)            %
%  where                                                                  %
%    X           Nxd matrix of parameter vectors                          %
%    n           lenght of simulation                                     %
%                                                                         %
% Please check the following papers                                       %
%  Massoud, E.C., J. Huisman, E. Benincà, M.C. Dietze, W. Bouten, and     % 
%      J.A. Vrugt (2018), Probing the limits of predictability: data      %
%      assimilation of chaotic dynamics in complex food webs, Ecology     %
%      Letters, 21 (1), pp 93-103, https://doi.org/10.1111/ele.12876      %
%  Benincà, E., K.D. Jöhnk, R. Heerkloss, and J. Huisman (2009), Coupled  %
%      predator–prey oscillations in a chaotic food web, Ecology Letters, %
%      12, pp. 1367-1378, doi:10.1111/j.1461-0248.2009.01391.x            %
%  Vandermeer, J. (1993), Loose coupling of predator-prey cycles:         %
%      entrainment, chaos, and intermittency in the classic MacArthur     %
%      consumer-resource equations, American Naturalist, 141, pp. 687-716 %
%  Vandermeer, J. (2004), Coupled oscillations in food webs: balancing    % 
%      competition and mutualism in simple ecological models. American    %
%      Naturalist, 163, pp. 857-867                                       %
%  Vandermeer, J. (2006), Oscillating populations and biodiversity        % 
%      maintenance, Bioscience, 56, pp. 967-975                           % 
%                                                                         %
%  © Written by Jasper A. Vrugt & Abdullah Sahin                          %
%    University of California Irvine                                      %
% %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% %

N = size(X,1);                                  % # parameter vectors
Y = nan(N,n,4);                                 % initialize model output 
                                                % N vectors, n simulated counts 
                                                % & k = 4 species

% Built-in ode45 function to solve 4 differential equations t = 0 to t_end
switch tbx                                      % N simulations parallel?
    case 0      % sequential
        prev = 0;                               % initialize ct printing
        for i = 1:N
            [~,u] = ode45(@(t,u) ...
                VDM(t,u,X(i,1:d),mV), ...
                (0:dt:t_end),u0,ode_options);
            Y(i,:,:) = reshape(u,1,n,4);        % Each species in own Y
            fprintf(1, repmat('\b',1,prev));    % print progress
            prev = fprintf(['VDM:MODEL ' ...
                'EVALUATIONS: %3.2f%% DONE'],100*(i/N));
        end
    case 1      % parallel evaluation
        poolobj = gcp;                          % Pool already open?
        if isempty(poolobj)                     % Open workers
            poolobj = parpool('local');
        end
        fprintf(['VDM: %d workers are used to evaluate the ' ...
            'vandermeer population model %d times\n'], ...
            poolobj.NumWorkers,N);
        parfor i = 1:N
            [~,u] = ode45(@(t,u) ...
                VDM(t,u,X(i,1:d),mV), ...
                (0:dt:t_end),u0,ode_options);
            Y(i,:,:) = reshape(u,1,n,4);        % Each species in own Y
            fprintf('.')                        % print progress    
        end
        poolobj = gcp('nocreate');              % delete the pool
        delete(poolobj);
end
[~,n,K] = size(Y); SS = nan(N,8,K); ct = 1;     % Summary statistics 
% # peaks tells us about # predator prey cycles and
%   std tells us about dynamics around mean
for k = 1:K                                     % Loop over each species
    for i = 1:N                                 % # peaks of simulation
        [mxL,mxM] = peakfinder(Y(i,1:n,k),0.2);
        if ~isempty(mxM)
            %peak = [mxL mxM];                  % 1*501 values for each i, there are one min values for each k
            %peak_cell{i,:} = num2cell(peak);   % each peak different dimensions?
            dim = size(mxM);                    
            mn_dis = n/(dim(2) + 1);            % mean distance peaks?
            peak_mn = mean(mxM);                % mean peak amplitude?
            peak_mx = max(mxM);                 % maximum peak
            % coherence between two curves??
            % quarter delay ??
            % m = 1: SD
            % m = 3: number of peaks
            % m = 4: average distance between each peaks
            % m = 5: the average amplitude of peaks
            % m = 6: the max of peak
            % for m = 7 - 9:
            % (1) case original:
            % m = 7: beta  = x(1)
            % m = 8: alpha = x(2)
            % (2) case new:
            % m = 7: beta  = X(1)
            % m = 8: alpha = [X(2), X(3)]
            % case original:
            SS(ct,1:8,k) = [std(Y(i,1:n,k),[],2), ...
                mean(Y(i,1:n,k),2),numel(mxL),...
                mn_dis,peak_mn,peak_mx,X(i,1),X(i,2)];
            ct = ct + 1;
        else
            % do nothing
        end
        % case 'new' % metrics + 1
        % X_avg(i) = (X(i,2) + X(i,3))/2
        %     SS(i,1:8,k) = [std(Y(i,1:n,k),[],2),mean(Y(i,1:n,k),2),numel(maxLoc)...
        %     avg_dis,peak_avg,peak_max,X(i,1),X_avg(i)];
    end
end

end

File: Python/example_25/VDM_model.py
import numpy as np
from scipy.integrate import odeint
import multiprocessing
import os

def VDM_model(X, n, d, u0, ode_options, dt, t_end, mV, tbx):
    """
    Simulates the counts and summary metrics of the Vandermeer predator-prey model.

    Args:
        X (ndarray): Nxd matrix of parameter vectors.
        n (int): Length of simulation.
        d (int): Number of parameters.
        u0 (list): Initial conditions for 4 species.
        ode_options (dict): Options for ODE solver.
        dt (float): Time step size.
        t_end (float): End time for simulation.
        mV (int): Model version.
        tbx (int): Whether to run in parallel (1 for parallel, 0 for sequential).

    Returns:
        X (ndarray): The parameter vectors.
        Y (ndarray): The simulated counts.
        SS (ndarray): Summary statistics for each simulation.
    """
VDM(t, u_in, par, parameterization, tbx):
    N = X.shape[0]  # Number of parameter vectors
    Y = np.full((N, n, 4), np.nan)  # Initialize model output (N simulations, n time steps, 4 species)

    if tbx == 0:  # Sequential computation
        prev = 0
        for i in range(N):
            t = np.arange(0, t_end + dt, dt)  # Time array from 0 to t_end with step dt
            sol = odeint(VDM, u0, t, args=(X[i, :d], mV), **ode_options)  # Solve ODE system
            Y[i, :, :] = sol  # Store solution
            prev = print_progress(i, N, prev)
    
    elif tbx == 1:  # Parallel computation
        pool = multiprocessing.Pool()  # Initialize multiprocessing pool
        print(f"VDM: {pool._processes} workers are used to evaluate the model {N} times")

        def simulate(i):
            t = np.arange(0, t_end + dt, dt)
            sol = odeint(VDM, u0, t, args=(X[i, :d], mV), **ode_options)
            return sol

        results = pool.map(simulate, range(N))  # Parallelize the simulations
        for i in range(N):
            Y[i, :, :] = results[i]  # Store results
        pool.close()
        pool.join()

    # Summary statistics computation
    _, n, K = Y.shape
    SS = np.full((N, 8, K), np.nan)
    ct = 0

    # Loop over each species
    for k in range(K):
        for i in range(N):
            mxL, mxM = peakfinder(Y[i, :, k], 0.2)  # Find peaks
            if mxM.size > 0:
                # Calculate summary statistics for this simulation
                dim = mxM.shape
                mn_dis = n / (dim[1] + 1)  # Mean distance between peaks
                peak_mn = np.mean(mxM)  # Mean peak amplitude
                peak_mx = np.max(mxM)  # Max peak value
                
                SS[ct, :, k] = [
                    np.std(Y[i, :, k]),  # Standard deviation
                    np.mean(Y[i, :, k]),  # Mean value
                    len(mxL),  # Number of peaks
                    mn_dis,  # Mean distance between peaks
                    peak_mn,  # Mean peak amplitude
                    peak_mx,  # Max peak
                    X[i, 0],  # First parameter (beta)
                    X[i, 1]  # Second parameter (alpha)
                ]
                ct += 1

    return X, Y, SS


def VDM(t, u_in, par, parameterization):
    """
    Coupled ordinary differential equations for the Vandermeer predator-prey model.
    This function defines the system of differential equations for the four species.
    """
    # Assign the states
    P1, P2, Z1, Z2 = u_in

    if parameterization == 1:
        beta = par[0]
        alfa1 = par[1]
        alfa2 = alfa1
        r1 = par[2]
        r2 = par[3]
        g = par[4]
        K1 = par[5]
        K2 = K1
        m1 = par[6]
        m2 = m1
        H = par[7]
    elif parameterization == 2:
        beta = par[0]
        alfa1 = par[1]
        alfa2 = par[2]
        r1 = par[3]
        r2 = par[4]
        g = par[5]
        K1 = par[6]
        K2 = par[7]
        m1 = par[8]
        m2 = par[9]
        H = par[10]
    else:
        raise ValueError("Unknown parameterization")

    # Define the system of ODEs
    dP1dt = r1 * P1 * (1 - (1 / K1) * (P1 + alfa1 * P2)) - (
        g * P1 * Z1 / (H + (P1 + beta * P2)) + g * beta * P1 * Z2 / (H + (beta * P1 + P2))
    )
    dP2dt = r2 * P2 * (1 - (1 / K2) * (alfa2 * P1 + P2)) - (
        g * beta * P2 * Z1 / (H + (P1 + beta * P2)) + g * P2 * Z2 / (H + (beta * P1 + P2))
    )
    dZ1dt = g * (P1 + beta * P2) * Z1 / (H + (P1 + beta * P2)) - m1 * Z1
    dZ2dt = g * (beta * P1 + P2) * Z2 / (H + (beta * P1 + P2)) - m2 * Z2

    # Return the system of differential equations as a vector
    return [dP1dt, dP2dt, dZ1dt, dZ2dt]


def peakfinder(signal, threshold):
    """
    Function to find peaks in the signal that are above the given threshold.
    Returns the indices of the peaks (mxL) and the peak values (mxM).
    """
    mxL = []
    mxM = []

    for i in range(1, len(signal) - 1):
        if signal[i] > signal[i - 1] and signal[i] > signal[i + 1] and signal[i] > threshold:
            mxL.append(i)
            mxM.append(signal[i])

    return np.array(mxL), np.array(mxM)


def print_progress(i, N, prev):
    """
    Helper function to print the progress of the simulations.
    """
    prev = print(f'VDM: MODEL EVALUATIONS: {100*(i+1)/N:.2f}% DONE', end='\r')
    return prev
